stdlib/asm/crypto/miden_vm_verifier/deep_queries.masm

```diff
@@ -0,0 +1,276 @@
+
+    exec.constants::get_fri_queries_address
+    # => [query_ptr, ...]
+
+    padw exec.constants::ood_fixed_term_horner_evaluations_ptr mem_loadw
+    # => [q_gz_1, q_gz_0, q_z_1, q_z_0, query_ptr, ...]
+    # => [W, query_ptr, ...] where W is [q_gz_1, q_gz_0, q_z_1, q_z_0]
+
+    # Get pointer to help test for the last query to be processed
+    exec.constants::fri_com_ptr
+    # => [query_end_ptr, W, query_ptr, ...]
+    dup.5
+    # => [query_ptr, query_end_ptr, W, query_ptr, ...]
+
+    # Store pointers to alpha, memory region to store trace row for current query and a fresh
+    # accumulator value i.e., (0, 0).
+    push.0.0
+    exec.constants::deep_rand_alpha_nd_ptr
+    exec.constants::current_trace_row_ptr
+    exec.constants::tmp2 mem_storew
+
+    # Compute the negations of z and gz where z is the OOD point
+    # We do it here as this computation is common to all queries.
+    exec.constants::z_ptr mem_loadw
+    # => [zN_1, zN_0, z1, z0, query_ptr, query_end_ptr, W, query_ptr, ...]
+    drop drop
+    neg swap neg  
+    # => [-z0, -z1, query_ptr, query_end_ptr, W, query_ptr, ...]
+    dup.1 exec.constants::get_trace_domain_generator mul
+    # => [-gz1, -z0, -z1, query_ptr, query_end_ptr, W, query_ptr, ...]
+    swap
+    # => [-z0, -gz1, -z1, query_ptr, query_end_ptr, W, query_ptr, ...]
+    dup exec.constants::get_trace_domain_generator mul
+    # => [-gz0, -z0, -gz1, -z1, query_ptr, query_end_ptr, W, query_ptr, ...]
+    swap
+    # => [-z0, -gz0, -gz1, -z1, query_ptr, query_end_ptr, W, query_ptr, ...]
+    # Save to temporary location `tmp1` for later use when computing the denominators
+    exec.constants::tmp1 mem_storew
+    # => [Y, query_ptr, query_end_ptr, W, query_ptr, ...]                   # (Cycles: 28)
+
+    push.1
+    while.true
+        # I)
+        #
+        # Load the (main, aux, constraint)-traces rows associated with the current query and get
+        # the index of the query.
+        #
+        # Cycles: 280
+        exec.load_query_row
+        #=>[Y, X, index, query_ptr, query_end_ptr, W, query_ptr, ...]
+
+        # II)
+        #
+        # Compute x := offset * domain_gen^index and denominators (x - z) and (x - gz)
+        #
+        # Cycles: 58
+        exec.deep_queries::compute_denominators
+        #=> [Z, X, x, index, query_ptr, query_end_ptr, W, query_ptr, ...] where Z := [-gz1, x - gz0, -z1, x - z0]
+
+        # III)
+        # Compute final result
+        #
+        # Cycles: 62
+        exec.deep_queries::divide_and_add
+        #=> [eval1, eval0, x, index, query_ptr, query_end_ptr, W, query_ptr, ...]
+
+        # IV)
+        #
+        # Store [poe, index, eval_1, eval_0] where poe := g^index = x / offset and prepare stack
+        # for next iteration.
+
+        ## a) Compute poe
+        ##
+        ## Cycles: 4
+        movup.3 movup.3
+        exec.constants::get_domain_offset_inv mul
+        #=> [poe, index, eval1, eval0, query_ptr, query_end_ptr, W, query_ptr, ...]
```

Comment:
Similar to the last comment, it seems like this is also generic functionality. Maybe we could encapsulate it in a single procedure - e.g., `deep_queries::process_query_row`.

Basically, I'm trying to reduce the amount of boilerplate code by moving generic pieces into the `std::crypto::stark` modules.

Reply:
totally agree, I have again factored the code out and put it in the generic verifier